File: js/sumPairs.js
/**
 * Returns an array of unique pairs of numbers from the array which sum to the given target.
 *
 * @param {Array} arr The array to be checked
 * @param {Number} target The target sum
 * @returns {Array} An array of pairs of numbers from the array which sum to the given target
 */
const sumPairs = (arr, target) => {

  // const result = new Set();

  // for (let i = 0; i < arr.length - 1; i++) {
  //   for (let j = i + 1; j < arr.length; j++) {
  //     const a = arr[i];
  //     const b = arr[j];

  //     if (a + b === target) {
  //       result.add([a, b]);
  //     }
  //   }
  // }

  // console.log(Array.from(result))

  // return Array.from(result);

  const result = [];

  for (let i = 0; i < arr.length - 1; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      const a = arr[i];
      const b = arr[j];


      if (a + b === target && !result.includes([a, b]) && !result.includes([b, a])) {
        result.push([a, b]);
      }
    }
  }

  return result;
};

module.exports = { sumPairs };

File: python/sum_pairs.py


def sum_pairs(arr, target):
    """
    This function returns all unique pairs of elements in the given array which sum to the target.

    Args:
        arr (list): The array of elements
        target (int): The target sum of pairs of elements 

    Returns:
        list[list[a, b]]: List of pairs of elements who sum to the target
    """

    range(start, stop, step)
    [1,2,3][1:5:2]
    [1,2,3][1:]

    result = set()

    for i, a in enumerate(arr[:-1]):
        for b in arr[i + 1:]:
            if a + b == target:
                result.add(frozenset(a, b))  # elements of a set must be hashable (immutable)
                                             # lists are mutable so we would use tuples
                                             # but we don't want duplicate pairs so we use
                                             # frozensets which are immutable sets

    return [list(item) for item in result]

    result = list()

    for i in range(len(arr) - 1):
        for j in range(i + 1, len(arr)):
            a, b = arr[i], arr[j]
            
            if a + b == target and [a, b] not in result and [b, a] not in result:
                result.append([a, b])

    return result


result = sum_pairs([1,2,3,4,5], 9) # [[4,5]]
print(result)
result = sum_pairs([1,2,3,4,5], 7) # [[2,5], [3,4]]
print(result)
result = sum_pairs([3, 1, 5, 8, 2], 27) # 'unable to find pairs'
print(result)
